### PR TITLE
OM3. As a member of an organization, I want to configure its settings, to keep it up to date.

### PR DESCRIPTION
**Pull request by meyginix**
https://trello.com/c/iX01kyIV/55-40-om3-as-a-member-of-an-organization-i-want-to-configure-its-settings-to-keep-it-up-to-date

### CODE REVIEW

File: app/imports/api/checkers.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { Meteor } from 'meteor/meteor';
+
+
+export const checkUserId = (userId, errMessage) => {
+  if (!userId) {
+    throw new Meteor.Error(403, errMessage);
+  }
```
Comment (trsdln) [inline app/imports/api/checkers.js]: @meyginix it doesn't make sense

File: app/imports/api/departments/methods.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  validate: IdSchema.validator(),
+
+  run(doc) {
+    checkUserId(
+      this.userId, 'Unauthorized user cannot remove a department'
+    );
+
+    return DepartmentService.remove(doc);
+  }
+});
```
Comment (trsdln) [inline app/imports/api/departments/methods.js]: @meyginix Do you have security checks functionality?

Comment (meyginix) [inline app/imports/api/departments/methods.js]: Not yet

File: app/imports/ui/components/dashboard/organization-settings/departments.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+Template.OrganizationSettings_Departments.viewmodel({
+  addDepartmentForm() {
+    Blaze.renderWithData(
+      Template.OrganizationSettings_Department,
+      {
+        onChange: this.onChangeCb(),
+        onDelete: this.onDeleteCb()
+      },
+      this.templateInstance.$('#departments-forms')[0]
+    );
```
Comment (trsdln) [inline app/imports/ui/components/dashboard/organization-settings/departments.js]: @meyginix is it normal practice to use blaze.renderwithdata within viewmodel package?

File: app/imports/ui/components/dashboard/organization-settings/departments.html [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+      </span>
+    </h4>
+  </div>
+  <div class="card-block-collapse collapse" id="departments">
+    <div class="card-block">
+      <div id="departments-forms">
+        {{#each departments}}
+          {{> OrganizationSettings_Department 
+              _id=this._id
+              name=this.name
```
Comment (trsdln) [inline app/imports/ui/components/dashboard/organization-settings/departments.html]: @meyginix this - redundant

Comment (trsdln) [inline app/imports/ui/components/dashboard/organization-settings/departments.html]: @meyginix probably pass whole department document is more reliable way.

File: app/imports/ui/components/dashboard/organization-settings/departments.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      });
+    }
+  },
+  onDelete(viewModel) {
+    if (!viewModel._id) {
+      Blaze.remove(viewModel.templateInstance.view);
+      return;
+    }
+
+    if (!confirm('Delete this department?')) return;
```
Comment (trsdln) [inline app/imports/ui/components/dashboard/organization-settings/departments.js]: @meyginix curly braces missing. 

@Praxie Probably we need to setup some linter for stuff like this.

File: app/imports/ui/components/dashboard/organization-settings/main-settings.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      this.setSavingState(false);  
+
+      if (err) {
+        toastr.error('Failed to update a name');
+      }
+    });
+  },
+  updateCurrency(currency) {
+    const current = this.currency();
+    if (currency === current) return;
```
Comment (trsdln) [inline app/imports/ui/components/dashboard/organization-settings/main-settings.js]: @meyginix  lines 17, 33 same problem

File: app/imports/ui/components/dashboard/organization-settings/departments.js (hunk 3 above)
Comment (trsdln) [inline app/imports/ui/components/dashboard/organization-settings/departments.js]: @meyginix @Praxie Better use some reactive data source (like reactive array or anonymous collection) to control it.

File: app/imports/ui/components/dashboard/organization-settings/nc-guidelines.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+    this.setSavingState(true);
+
+    setGuideline.call({ _id, ncType, text }, (err) => {
+      this.setSavingState(false);
+
+      if (err) {
+        toastr.error('Failed to update non-conformity guidelines');
+      }
+    });
```
Comment (trsdln) [inline app/imports/ui/components/dashboard/organization-settings/nc-guidelines.js]: @meyginix @Praxie probably we need some centralized solution for error handling. Consider something like this:
```js
 handleMethodResult: function (onSuccess) {
    return function (err, res) {
      if (err) {
        toastr.error(err);
      } else {
        if (_.isFunction(onSuccess)) {
          onSuccess(res);
        }
      }
    };
  }
```

File: app/imports/ui/components/dashboard/organization-settings/settings.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import { Departments } from '/imports/api/departments/departments.js';
+import { StandardsTypes } from '/imports/api/standards-types/standards-types.js';
+import {
+  StandardsBookSections
+} from '/imports/api/standards-book-sections/standards-book-sections.js';
+
+
+Template.OrganizationSettings.viewmodel({
+  organization() {
+    const serialNumber = Number(FlowRouter.getParam('orgSerialNumber'));
```
Comment (trsdln) [inline app/imports/ui/components/dashboard/organization-settings/settings.js]: @meyginix it doesn't make sense. parseInt instead.

**Comment by trsdln** [comment]
@Praxie I finished reviewing.

File: app/imports/ui/components/dashboard/organization-settings/departments.html (hunk 4 above)
Comment (meyginix) [inline app/imports/ui/components/dashboard/organization-settings/departments.html]: When we pass name separately, we automatically have ```name``` property in viewmodel that is reactive, can be easily bound to UI control and easily retrieved/changed. Passing whole department requires more additional code to make viewmodel work

Comment (trsdln) [inline app/imports/ui/components/dashboard/organization-settings/departments.html]: @meyginix OK. Thanks for explanation.

File: app/imports/api/departments/methods.js (hunk 2 above)
Comment (trsdln) [inline app/imports/api/departments/methods.js]: @Praxie it should be added asap. Adding all this checks to existing code is pretty annoying.

File: app/imports/ui/components/dashboard/organization-settings/nc-guidelines.js (hunk 7 above)
Comment (Praxie) [inline app/imports/ui/components/dashboard/organization-settings/nc-guidelines.js]: @trsdln it looks good!

File: app/imports/ui/components/modals/modal-heading.js [hunk 9]
```diff
@@ -1,5 +1,11 @@
 Template.ModalHeading.viewmodel({
   isSaving: false,
+  toggleSaving() {
+    this.isSaving(true);
+    setTimeout(() => {
```
Comment (trsdln) [inline app/imports/ui/components/modals/modal-heading.js]: @Praxie Meteor.setTimeout

**Pull request closed without merging**